File: LeetCode/Easy/589 - N-ary Tree Preorder Traversal/589 - N-ary Tree Preorder Traversal.js
/**
 * // Definition for a Node.
 * function Node(val, children) {
 *    this.val = val;
 *    this.children = children;
 * };
 */

/**
 * @param {Node|null} root
 * @return {number[]}
 */
 var preorder = function(root) {
    const ret = [];
    
    const solve = (node) => {
        if (!node) return;
        ret.push(node.val);
        
        for (const child of node.children) {
            solve(child);
        }
    };
    solve(root);
    
    return ret;
};

File: LeetCode/Easy/589 - N-ary Tree Preorder Traversal/589 - N-ary Tree Preorder Traversal.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
public:
    vector<int> preorder(Node* root) {
        vector<int> ret;
        preorder(root, ret);
        return ret;
    }
    
    void preorder(Node* root, vector<int> &vc) {
        if (!root) return;
        vc.push_back(root->val);
        
        for (auto child : root->children) {
            preorder(child, vc);
        }
    }
};